// gym-js/dist/env.js
"use strict";
// Environments
// ===============================================
Object.defineProperty(exports, "__esModule", { value: true });
// Toy text envs
var FrozenLake_1 = require("./envs/text/FrozenLake");
exports.FrozenLake = FrozenLake_1.default;
// -----------------------------------------------
// Algorithmic envs
var Copy_1 = require("./envs/algorithmic/Copy");
exports.Copy = Copy_1.default;
var Reverse_1 = require("./envs/algorithmic/Reverse");
exports.Reverse = Reverse_1.default;
// -----------------------------------------------
// Arcade envs
var Snake_1 = require("./envs/arcade/snake/Snake");
exports.Snake = Snake_1.default;
// -----------------------------------------------
// ===============================================
// Core
var core_1 = require("./core");
exports.Wrapper = core_1.Wrapper;
exports.ObservationWrapper = core_1.ObservationWrapper;
exports.ActionWrapper = core_1.ActionWrapper;
exports.RewardWrapper = core_1.RewardWrapper;
var discrete_1 = require("./spaces/discrete");
exports.Discrete = discrete_1.default;
exports.DiscreteTuple = discrete_1.DiscreteTuple;
// -----------------------------------------------
